File: app/src/main/java/fr/paug/androidmakers/util/MapUtil.java
package fr.paug.androidmakers.util;

import android.text.TextUtils;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import fr.paug.androidmakers.model.Partners;
import fr.paug.androidmakers.model.Ribbon;
import fr.paug.androidmakers.model.SocialNetworkHandle;

public final class MapUtil {

    private MapUtil() {
        // no instance
    }

    public static int getInt(Map map, String key, int defaultValue) {
        return getInt(map.get(key), defaultValue);
    }

    public static int getInt(Object object, int defaultValue) {
        if (object instanceof Integer) {
            return (Integer) object;
        } else if (object instanceof Long) {
            return ((Long) object).intValue();
        } else {
            return defaultValue;
        }
    }

    public static String getString(Map map, String key) {
        Object object = map.get(key);
        if (object instanceof String) {
            return (String) object;
        } else {
            return null;
        }
    }

    public static List<SocialNetworkHandle> getSocialItems(Map map, String key) {
        final Object object = map.get(key);
        if (object instanceof List) {
            List<SocialNetworkHandle> socialNetworkHandles = null;
            for (Object socialItem : ((List) object)) {
                if (socialItem instanceof Map) {
                    final String name = getString((Map) socialItem, "name");
                    final String link = getString((Map) socialItem, "link");
                    if (!TextUtils.isEmpty(name) && !TextUtils.isEmpty(link)) {
                        if (socialNetworkHandles == null) {
                            socialNetworkHandles = new ArrayList<>();
                        }
                        socialNetworkHandles.add(new SocialNetworkHandle(name, link));
                    }
                }
            }
            return socialNetworkHandles;
        } else {
            return null;
        }
    }

    public static List<Ribbon> getRibbonItems(Map map, String key) {
        final Object object = map.get(key);
        if (object instanceof List) {
            List<Ribbon> ribbonList = null;
            for (Object socialItem : ((List) object)) {
                if (socialItem instanceof Map) {
                    final String name = getString((Map) socialItem, "abbr");
                    final String title = getString((Map) socialItem, "title");
                    final String link = getString((Map) socialItem, "url");
                    if (!TextUtils.isEmpty(name) && !TextUtils.isEmpty(link)) {
                        if (ribbonList == null) {
                            ribbonList = new ArrayList<>();
                        }
                        ribbonList.add(new Ribbon(name, title, link));
                    }
                }
            }
            return ribbonList;
        } else {
            return null;
        }
    }

    public static List<Partners> getPartnerList(Map map, String key) {
        final Object object = map.get(key);
        if (object instanceof List) {
            List<Partners> partnersList = null;
            for (Object partner : ((List) object)) {
                if (partner instanceof Map) {
                    final String name = getString((Map) partner, "name");
                    final String link = getString((Map) partner, "link");
                    final String imageUrl = getString((Map) partner, "imageUrl");
                    final String description = getString((Map) partner, "description");
                    if (!TextUtils.isEmpty(name) && !TextUtils.isEmpty(imageUrl)) {
                        if (partnersList == null) {
                            partnersList = new ArrayList<>();
                        }
                        partnersList.add(new Partners(name, imageUrl, link, description));
                    }
                }
            }
            return partnersList;
        } else {
            return null;
        }
    }

    public static int[] getIntArray(Map map, String key) {
        Object object = map.get(key);
        if (object instanceof List) {
            List list = (List) object;
            int[] result = new int[list.size()];
            for (int i = 0; i < list.size(); i++) {
                int value = getInt(list.get(i), -1);
                if (value < 0) {
                    return null;
                }
                result[i] = value;
            }
            return result;
        } else {
            return null;
        }
    }

}

File: app/src/main/java/fr/paug/androidmakers/util/MapUtil.kt
package fr.paug.androidmakers.util

import fr.androidmakers.store.model.Partners
import fr.androidmakers.store.model.Ribbon
import fr.androidmakers.store.model.SocialNetworkHandle
import java.util.*

object MapUtil {

    @JvmStatic
    fun Map<*, *>.getInt(key: String, defaultValue: Int): Int {
        return (this[key] as? Number)?.toInt() ?: defaultValue
    }

    @JvmStatic
    fun Map<*, *>.getString(key: String): String? {
        return this[key] as? String
    }

    @JvmStatic
    fun getSocialItems(map: Map<Any?, Any?>, key: String): List<SocialNetworkHandle>? {
        return (map[key] as? List<*>)
                ?.let { list ->
                    val socialNetworkHandles: MutableList<SocialNetworkHandle> = ArrayList()
                    list.forEach { item ->
                        (item as? Map<*, *>)?.apply {
                            val name = this.getString("name")
                            val link = this.getString("link")
                            if (name.isNullOrBlank().not() && link.isNullOrBlank().not()) {
                                socialNetworkHandles.add(SocialNetworkHandle(name, link))
                            }
                        }
                    }
                    return socialNetworkHandles.takeIf { it.size > 0 }
                }
    }

    @JvmStatic
    fun getRibbonItems(map: Map<*, *>, key: String): List<Ribbon>? {
        return (map[key] as? List<*>)
                ?.let { list ->
                    val ribbonList: MutableList<Ribbon> = ArrayList()
                    list.forEach { item ->
                        (item as? Map<*, *>)?.apply {
                            val name = this.getString("abbr")
                            val title = this.getString("title")
                            val link = this.getString("url")
                            if (name.isNullOrBlank().not() && link.isNullOrBlank().not()) {
                                ribbonList.add(Ribbon(name, title, link))
                            }
                        }
                    }
                    return ribbonList.takeIf { it.size > 0 }
                }
    }

    @JvmStatic
    fun getPartnerList(map: Map<*, *>, key: String): List<Partners>? {
        return (map[key] as? List<*>)
                ?.let { list ->
                    val partnersList: MutableList<Partners> = ArrayList()
                    list.forEach { item ->
                        (item as? Map<*, *>)?.apply {
                            val name = this.getString("name")
                            val link = this.getString("link")
                            val imageUrl = this.getString("imageUrl")
                            val description = this.getString("description")
                            if (name.isNullOrBlank().not() && imageUrl.isNullOrBlank().not()) {
                                partnersList.add(Partners(name, imageUrl, link, description))
                            }
                        }
                    }
                    return partnersList.takeIf { it.size > 0 }
                }
    }

    @JvmStatic
    fun getIntArray(map: Map<*, *>, key: String): IntArray? {
        return (map[key] as? List<*>)
                ?.let { list ->
                    val result = IntArray(list.size)
                    list.forEachIndexed { index, item ->
                        val value = (item as? Number)?.toInt() ?: -1
                        if (value < 0) {
                            return null
                        }
                        result[index] = value
                    }
                    return result
                }
    }

}